cmake/modules/kconfig.cmake

```diff
@@ -111,6 +111,17 @@ string(REPLACE ";" "\\\;" SHIELD_AS_LIST_ESCAPED "${SHIELD_AS_LIST}")
 # cmake commands are escaped differently
 string(REPLACE ";" "\\;" SHIELD_AS_LIST_ESCAPED_COMMAND "${SHIELD_AS_LIST}")
 
+# A Kconfig.board indicates legacy style board definition, whereas
+# Kconfig.${BOARD} indicates new and improved style.
+if(EXISTS ${BOARD_DIR}/Kconfig.board AND NOT EXISTS ${BOARD_DIR}/Kconfig.${BOARD})
+  set(BOARD_SCHEME v1)
+elseif(NOT EXISTS ${BOARD_DIR}/Kconfig.board AND EXISTS ${BOARD_DIR}/Kconfig.${BOARD})
+  set(BOARD_SCHEME v2)
+elseif(EXISTS ${BOARD_DIR}/Kconfig.board AND EXISTS ${BOARD_DIR}/Kconfig.${BOARD})
+  message(WARNING "Mixed board scheme (v1 + v2) not allowed, using board scheme v2.")
+  set(BOARD_SCHEME v2)
+endif()
+
```

Comment:
In https://github.com/zephyrproject-rtos/zephyr/pull/52272, I make an argument that we should really be using the devicetree to determine the board. The kconfig style was added before we had devicetree support. Then

- devicetree was introduced
- we started running it even before kconfig
- it is the **hardware** description, after all

So I think it makes more sense to do board detection from devicetree files, if we're going to continue to use file system conventions.

As an alternative to using file system conventions, we could use the board YAML file to include the metadata we need, and run a python script that generates the cmake settings we want based on that. This is more akin to the approach in the snippets RFC (https://github.com/zephyrproject-rtos/zephyr/pull/53640). I personally think a config file based approach is more flexible and future-proof, but I'd be curious to know if you think otherwise.

Reply:
> In https://github.com/zephyrproject-rtos/zephyr/pull/52272, I make an argument that we should really be using the devicetree to determine the board.

sure, I have no problem with that, and it sounds like a better approach.
But all I had at the time was a hammer, so therefore the approach.

>  we could use the board YAML file to include the metadata we need, and run a python script that generates the cmake settings we want based on that. 

Could be an option, but today we don't use the board yaml in CMake and afaik it's only used by twister today.
Meaning, the content of this file bears the risk of being messy if trying to cover too different use-cases / tools.

But at this point I believe it to be very valuable to have as many good ideas posted.
Will help to pick the most suitable path forward.